_posts/2023-01-14-trivially-relocatable.md

```diff
@@ -0,0 +1,656 @@
+of `T` becomes 4, which is the size of its member. There is no extra byte for its base class `S`:
+
+```cpp
+struct S {};
+struct T : public S { int32_t a; };
+static_assert(sizeof(T) == 4);
+```
+
+This presents the first case where the data size of an object matters. If we were to
+`memcpy(&s, &from, sizeof(S))` into an `S*` but it so happens that the object is a `T`, we would
+overwrite one byte of `T::a` with garbage!
+
+The [Clang](https://godbolt.org/z/oejGWP694) and [GCC](https://godbolt.org/z/jjxcMYd3n) (but not
+[MSVC 19](https://godbolt.org/z/7fczM93Gn)) compilers have taken this further, and will generally
+make use of the tail padding in any base class that is not a [Standard-Layout](
+https://en.cppreference.com/w/cpp/language/classes#Standard-layout_class) type.
+
+Recall our earlier example of a struct with tail padding. We mark the `b` member as `private` in
+order to make the type not [Standard-Layout](
+https://en.cppreference.com/w/cpp/language/classes#Standard-layout_class):
+
+```cpp
+struct S {
+    int32_t a;
+  private:
+    int8_t b;
+    // 3 bytes of tail padding.
+};
+static_assert(sizeof(S) == 8);
+```
+
+Then if we make a subclass of `S`, the compiler is entitled to place members into the tail padding
+of `S`, as we see in `T` below:
+
+```cpp
+struct T : S {
+    int8_t c;
+    // 2 bytes of tail padding.
+};
+static_assert(sizeof(S) == 8);
+static_assert(sizeof(T) == 8);
+```
+
+The size of `T` is the same as `S` because the member `c` has been placed inside the tail padding of
+the base class `S`. If we were to `memcpy(&s, &from, sizeof(S))` into an `S*` that has been
+subclassed by `T`, we would overwrite one byte of `T::c` with garbage! However if we
+`memcpy(&s, &from, sus::mem::data_size_of<S>())`, then we copy only 5 bytes (the `int32_t a`
+and `int8_t b`) into `s` and avoid clobbering the members of `T` or any other subclass in its tail
+padding.
+
+##### The `[[no_unique_address]]` attribute
+
+C++20 introduces the `[[no_unique_address]]` attribute which can appear on a class member
+declaration. It tells the compiler to allow the member's tail padding to be used by later
+members.
+
+This new attribute allows us to compose a new type with a member `S` and which makes use of the tail
+padding in `S`. Here we use the same non-Standard-Layout type `S` from the previous section.
+We add another member below `S` but the size of `T`
+[is again no larger than S](https://godbolt.org/z/GhbvTaecM) because the
+member `T::c` has been located in the tail padding of `S`.
+
+```cpp
+struct T {
+    [[no_unique_address]] S s;
+    int8_t c;
+    // 2 bytes of tail padding.
+};
+static_assert(sizeof(S) == 8);
+static_assert(sizeof(T) == 8);
+```
+
+The side effect of the `[[no_unique_address]]` attribute is that `memcpy()` can do the wrong thing
+again. If we `memcpy(&s, &from, sizeof(S))` into an `S*` that points to the member `T::s` of `T`,

```

Comment:
memcpy(&t.s, &s, ...)?

Or "into an S that is actually the embedded `T::s` field, …"?

Reply:
Thanks, "into an `S` that is the embedded `T::s` member"